File: aspnetcore/MISA.WEB05.CEGOV/MISA.WEB05.CEGOV.Application/Services/AwardService.cs
using AutoMapper;
using Microsoft.AspNetCore.Http;
using MISA.WEB05.CEGOV.Domain;

namespace MISA.WEB05.CEGOV.Application
{
    public class AwardService : BaseService<Award, AwardDto, AwardCreateDto, AwardUpdateDto>, IAwardService
    {
        #region Fields
        private readonly IAwardRepository _awardRepository;
        private readonly IAwardManager _awardManager;
        private readonly IExcelWorker _excelWorker;
        #endregion

        #region Constructor
        public AwardService(IAwardRepository awardRepository, IMapper mapper, IAwardManager awardManager, IExcelWorker excelWorker) : base(awardRepository, mapper)
        {
            _awardRepository = awardRepository;
            _awardManager = awardManager;
            _excelWorker = excelWorker;
        }
        #endregion

        #region Methods
        /// <summary>
        /// Ánh xạ AwardCreateDto sang Award
        /// </summary>
        /// <param name="awardCreateDto">Đối tượng AwardCreateDto</param>
        /// <returns>Đối tượng Award</returns>
        /// Created by: ntlong ( 23/07/2023 )
        public override async Task<Award> MapEntityCreateDtoToEntity(AwardCreateDto awardCreateDto)
        {
            // Check mã danh hiệu đã tồn tại chưa
            await _awardManager.CheckAwardExistByCode(awardCreateDto.AwardCode);
            // Map sang Award
            var award = new Award()
            {
                AwardId = Guid.NewGuid(),
                AwardName = awardCreateDto.AwardName,
                AwardCode = awardCreateDto.AwardCode,
                AwardObject = awardCreateDto.AwardObject,
                AwardLevel = awardCreateDto.AwardLevel,
                AwardType = awardCreateDto.AwardType,
                AwardStatus = awardCreateDto.AwardStatus,
                IsSystemData = false,
                CreatedDate = DateTime.Now,
                CreatedBy = null,
                ModifiedDate = null,
                ModifiedBy = null,
            };
            return award;
        }

        /// <summary>
        /// Ánh xạ AwardUpdateDto sang Award
        /// </summary>
        /// <param name="awardUpdateDto">Đối tượng AwardUpdateDto</param>
        /// <returns>Đối tượng Award</returns>
        /// Created by: ntlong ( 23/07/2023 )
        public override async Task<Award> MapEntityUpdateDtoToEntity(AwardUpdateDto awardUpdateDto)
        {
            // Lấy ra mã cũ để check xem mã mới có bị trùng với các mã khác ngoài mã cũ không
            var oldAward = await _awardRepository.GetByIdAsync(awardUpdateDto.AwardId);
            var oldCode = oldAward.AwardCode;

            // Check mã danh hiệu đã tồn tại chưa
            await _awardManager.CheckAwardExistByCode(awardUpdateDto.AwardCode, oldCode);

            // Map sang Award
            var award = new Award()
            {
                AwardId = awardUpdateDto.AwardId,
                AwardName = awardUpdateDto.AwardName,
                AwardCode = awardUpdateDto.AwardCode,
                AwardObject = awardUpdateDto.AwardObject,
                AwardLevel = awardUpdateDto.AwardLevel,
                AwardType = awardUpdateDto.AwardType,
                AwardStatus = awardUpdateDto.AwardStatus,
                IsSystemData = false,
                ModifiedDate = DateTime.Now,
                ModifiedBy = null,
            };
            return award;
        }

        /// <summary>
        /// Ánh xạ AwardDto sang AwardDisplayDto
        /// </summary>
        /// <param name="awardDto">Đối tượng AwardUpdateDto</param>
        /// <returns>Đối tượng AwardDisplayDto</returns>
        /// Created by: ntlong ( 23/07/2023 )
        public AwardDisplayDto MapEntityDtoToEntityDisplayDto(AwardDto awardDto)
        {
            // Lấy các prop của 2 đối tượng để map giá trị
            var props = typeof(AwardDto).GetProperties();
            var displayProps = typeof(AwardDisplayDto).GetProperties();

            var displayDto = new AwardDisplayDto();

            for (int i = 0; i < props.Length; i++)
            {
                // Lấy ra từng giá trị của đối tượng AwardDto
                var value = props[i].GetValue(awardDto);

                // Nếu giá trị của prop nào của AwardDto có kiểu Enum thì map sang String
                if (value != null && value is Enum enumValue)
                {
                    value = enumValue.GetDescription();
                }
                // Lấy giá trị kiểu String vừa map để set cho đối tượng hiển thị
                displayProps[i].SetValue(displayDto, value);
            }

            return displayDto;
        }

        /// <summary>
        /// Sửa trường AwardStatus ở nhiều bản ghi
        /// </summary>
        /// <param name="ids">Mảng các id của các bản ghi cần sửa</param>
        /// <param name="newStatus">Giá trị mới của AwardStatus</param>
        /// Created by: ntlong ( 19/07/2023 )
        public async Task UpdateManyStatusAsync(List<Guid> ids, int newStatus)
        {
            await _awardRepository.UpdateManyStatusAsync(ids, newStatus);
        }


        /// <summary>
        /// Lấy danh sách bản ghi bao gồm phân trang, lọc, tìm kiếm, sắp xếp
        /// </summary>
        /// <param name="page">Số thứ tự của trang cần lấy</param>
        /// <param name="size">Số lượng tối đa một trang</param>
        /// <param name="search">Chuỗi tìm kiếm</param>
        /// <param name="searchFields">Danh sách các trường cần tìm kiếm</param>
        /// <param name="filteredObject">Giá trị lọc của đối tượng khen thưởng</param>
        /// <param name="filteredLevel">Giá trị lọc của cấp khen thưởng</param>
        /// <param name="filteredType">Giá trị lọc của loại khen thưởng</param>
        /// <param name="filteredStatus">Giá trị lọc của trạng thái</param>
        /// <param name="sortField">Trường được sắp xếp</param>
        /// <param name="sortType">Kiểu sắp xếp: true là tăng dần, false là giảm dần</param>
        /// <returns>Tổng số và danh sách bản ghi theo trang cần hiển thị</returns>
        /// Created by: ntlong ( 25/07/2023 )
        public async Task<(int Total, List<AwardDto> Data)> GetListFPSSAsync(int? page = null, int? size = null, string? search = null, List<string>? searchFields = null, AwardObjectEnum? filteredObject = null, AwardLevelEnum? filteredLevel = null, AwardTypeEnum? filteredType = null, AwardStatusEnum? filteredStatus = null, string? sortField = null, bool? sortType = null)
        {
            // Kiểm tra tính hợp lệ của page và size
            if (page <= 0 || size <= 0)
            {
                throw new ValidateException("Page hoặc PageSize nhỏ hơn 1");
            }

            // Nếu các trường lọc đều null thì filter = null
            Dictionary<string, object?>? filter;
            if (filteredObject == null && filteredLevel == null && filteredType == null && filteredStatus == null)
            {
                filter = null;
            }
            else // Ngược lại: Tạo một dictionary là các trường được lọc               
            {
                filter = new Dictionary<string, object?>
                {
                    { "AwardObject", filteredObject},
                    { "AwardLevel", filteredLevel},
                    { "AwardType", filteredType},
                    { "AwardStatus", filteredStatus}
                };
            }

            var (total, entities) = await _awardRepository.GetListFPSSAsync(page, size, search, searchFields, filter, sortField, sortType);
            var entityDtos = _mapper.Map<List<AwardDto>>(entities);

            //var data = new List<AwardDisplayDto>();
            //foreach ( var dto in entityDtos )
            //{
            //    var displayDto = MapEntityDtoToEntityDisplayDto(dto);
            //    data.Add(displayDto );
            //}

            //return (total, data);

            return (total, entityDtos);
        }

        /// <summary>
        /// Xuất dữ liệu ra file excel
        /// </summary>
        /// <returns></returns>
        /// Created by: ntlong ( 15/08/2023 )
        public async Task<byte[]> ExportToExcel(TableFormat tableFormat)
        {
            // Lấy dữ liệu trong database
            var awards = await _awardRepository.GetAllAsync();

            // Chuyển đổi thành Dto ẩn các trường dữ liệu thừa
            var dtos = _mapper.Map<IEnumerable<AwardDto>>(awards);

            var list = new List<IDictionary<string, object?>>();

            // Mapping dữ liệu
            foreach (var dto in dtos)
            {
                var type = dto.GetType();
                var props = type.GetProperties();
                var dic = props.ToDictionary(p => p.Name, p => p.GetValue(dto));
                foreach (var key in dic.Keys)
                {
                    var value = dic[key];
                    if (value != null && value is Enum enumValue)
                    {
                        var content = enumValue.GetDescription();
                        dic[key] = content;
                    }
                }
                list.Add(dic);
            }

            var file = await _excelWorker.ExportToExcel(tableFormat, list);
            return file;
        }

        /// <summary>
        /// Lấy ra Lấy thông tin và kiểm tra tính hợp lệ của file
        /// </summary>
        /// <param name="file">File được gửi từ client</param>
        /// Created by: ntlong ( 15/08/2023 )
        public async Task<ImportFileInfo> GetFileInfo(IFormFile file)
        {
            var fileInfo = await _excelWorker.GetFileInfo(file);
            return fileInfo;
        }

        /// <summary>
        /// Lấy ra số bản ghi hợp lệ và không hợp lệ của file được upload
        /// </summary>
        /// <param name="file">File được upload</param>
        /// <param name="sheetIndex">Thứ tự của sheet cần check</param>
        /// <param name="titleLine">Thứ tự của dòng tiêu đề</param>
        /// <returns></returns>
        /// Created by: ntlong ( 15/08/2023 )
        public async Task<(ImportDataStatistics DataStatistics, List<Dictionary<string, object?>> ValidList)> CheckImportFile(IFormFile file, int sheetIndex, int titleLine)
        {
            var fieldNames = new List<string>()
            {
                "AwardName", "AwardCode", "AwardObject", "AwardLevel", "AwardType", "AwardStatus"
            };
            var list = await _excelWorker.ReadImportFile(file, sheetIndex, titleLine, fieldNames);

            // Biến đếm số lượng bản ghi hợp lệ
            int validCount = 0;

            // Tạo list lưu những dictionary tương ứng bản ghi hợp lệ
            var validList = new List<Dictionary<string, object?>>();

            foreach (var dic in list)
            {
                // Kiểm tra xem bản ghi có hợp lệ không
                bool isValidRecord = await _awardManager.CheckImportRecordValidity(dic);
                if (isValidRecord) // Hợp lệ
                {
                    validCount++;
                    validList.Add(dic);
                }
            }

            var obj = new ImportDataStatistics
            {
                Total = list.Count,
                ValidNumber = validCount,
                InvalidNumber = list.Count - validCount
            };

            return (obj, validList);
        }

        /// <summary>
        /// Nhập dữ liệu từ file vào database
        /// </summary>
        /// <param name="file">File được upload</param>
        /// <param name="sheetIndex">Thứ tự của sheet cần check</param>
        /// <param name="titleLine">Thứ tự của dòng tiêu đề</param>
        /// <returns></returns>
        /// Created by: ntlong ( 15/08/2023 )
        public async Task ImportExcel(IFormFile file, int sheetIndex, int titleLine)
        {
            // Lấy ra list các bản ghi hợp lệ đã được kiểm tra
            var (_, validList) = await CheckImportFile(file, sheetIndex, titleLine);

            // Tạo list để map sang kiểu Award để insert vào db
            var dtos = new List<Award>();

            foreach (var dic in validList)
            {
                var createDto = new AwardCreateDto();
                var props = typeof(AwardCreateDto).GetProperties();
                foreach (var prop in props)
                {
                    if (dic.TryGetValue(prop.Name, out object? value))
                    {
                        if (prop.PropertyType.IsAssignableTo(typeof(Enum)))
                        {
                            var enumType = prop.PropertyType.Name;

                            value = _awardManager.ConvertTextToEnum(enumType, value as string);
                        }
                        prop.SetValue(createDto, value);
                    }
                }
                var dto = await MapEntityCreateDtoToEntity(createDto);
                dtos.Add(dto);
            }

            await _awardRepository.InsertManyAsync(dtos);
        }
        #endregion
    }
}

File: misa_w5/src/services/category/awardService.js
import request from "@/utils/request.js";
import store from "@/store/index";
import MISAEnum from "@/helpers/enum.js";
import MISAResource from "@/helpers/resource.js";

export default {
  /**
   * Lấy ra tất cả danh hiệu trong bảng award
   * Created by: ntlong ( 18/07/2023 )
   */
  async getAwards() {
    try {
      const res = await request.get("/v1/awards");
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /// <summary>
  /// Lấy danh sách bản ghi bao gồm phân trang, lọc, tìm kiếm, sắp xếp
  /// Ý nghĩa "FPSS": Filter-Paging-Search-Sort
  /// </summary>
  /// @param name="page": Số thứ tự của trang cần lấy
  /// @param name="size": Số lượng tối đa một trang
  /// @param name="search": Chuỗi tìm kiếm
  /// @param name="searchFields": Danh sách các trường cần tìm kiếm
  /// @param name="filteredObject": Giá trị lọc của đối tượng khen thưởng
  /// @param name="filteredLevel": Giá trị lọc của cấp khen thưởng
  /// @param name="filteredType": Giá trị lọc của loại khen thưởng
  /// @param name="filteredStatus": Giá trị lọc của trạng thái
  /// @param name="sortField": Trường được sắp xếp
  /// @param name="sortType": Kiểu sắp xếp: true là tăng dần, false là giảm dần
  /// <returns>Tổng số và danh sách bản ghi theo trang cần hiển thị</returns>
  /// Created by: ntlong ( 25/07/2023 )
  async getFPSSAwards(state) {
    try {
      const res = await request.get("/v1/Awards/ListFPSS", {
        params: {
          page: state.currentPage,
          size: state.pageSize,
          search: state.searchText,
          searchFields: ["awardName", "awardCode"],
          filteredObject: state.filter?.filteredObject,
          filteredLevel: state.filter?.filteredLevel,
          filteredType: state.filter?.filteredType,
          filteredStatus: state.filter?.filteredStatus,
          sortField: state.sortField ?? "AwardStatus",
          sortType: state.sortType ?? true,
        },
      });
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Thêm 1 danh hiệu vào trong bảng award
   * @param newAward: Đối tượng danh hiệu cần thêm mới
   * Created by: ntlong ( 18/07/2023 )
   */
  async insertOneAward(newAward) {
    try {
      const res = await request.post("/v1/Awards", newAward);
      store.commit("awardStore/closeForm"); // Đóng form
      store.commit("setShowToast", true); // Thông báo Thành Công
      return res.data;
    } catch (data) {
      if (data?.ErrorCode === 409) {
        // Show dialog báo lỗi
        store.commit("awardStore/setErrorAward", newAward);
        store.commit("setShowDialog", true);
        store.commit("setDialogType", MISAResource.dialog.type.error901);
        return;
      }
      if (data) {
        // Show dialog báo lỗi
        store.commit("setDialogType", MISAResource.dialog.type.error);
        store.commit("setError", data.UserMessage);
        store.commit("setShowDialog", true);
      }
    }
  },

  /**
   * Sửa 1 danh hiệu
   * @param newAward: Đối tượng danh hiệu cần sửa
   * Created by: ntlong ( 18/07/2023 )
   */
  async updateOneAward(newAward) {
    try {
      const res = await request.put(`/v1/Awards/${newAward.awardId}`, newAward);
      store.commit("awardStore/closeForm"); // Đóng form
      store.commit("setShowToast", true); // Thông báo Thành Công
      return res.data;
    } catch (data) {
      if (data?.ErrorCode === 409) {
        // Show dialog báo lỗi
        store.commit("awardStore/setErrorAward", newAward);
        store.commit("setDialogType", MISAResource.dialog.type.error901);
        store.commit("setShowDialog", true);
        return;
      }
      if (data) {
        // Show dialog báo lỗi
        store.commit("setDialogType", MISAResource.dialog.type.error);
        store.commit("setError", data.UserMessage);
        store.commit("setShowDialog", true);
      }
    }
  },

  /**
   * Xóa 1 danh hiệu
   * @param id: Id của danh hiệu cần xóa
   * Created by: ntlong ( 18/07/2023 )
   */
  async deleteOneAward(id) {
    try {
      const res = await request.delete(`/v1/Awards/${id}`);
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Xóa nhiều danh hiệu
   * @param listIds: Mảng các id của danh hiệu cần xóa
   * Created by: ntlong ( 18/07/2023 )
   */
  async deleteManyAward(listIds) {
    try {
      await request.delete(`/v1/Awards`, {
        headers: {
          "Content-Type": "application/json",
        },
        data: listIds,
      });
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Sửa trạng thái của nhiều danh hiệu
   * @param listIds: Mảng các id của danh hiệu cần sửa
   * @param newAwardStatus: trạng thái mới
   * Created by: ntlong ( 18/07/2023 )
   */
  async updateManyAwardStatus(listIds, newAwardStatus) {
    try {
      await request.put(`/v1/Awards/Status`, listIds, {
        headers: {
          "Content-Type": "application/json",
        },
        params: { newStatus: newAwardStatus },
      });
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Sửa trạng thái của nhiều danh hiệu
   * @param listIds: Mảng các id của danh hiệu cần sửa
   * @param newAwardStatus: trạng thái mới
   * Created by: ntlong ( 18/07/2023 )
   */
  async exportExcel() {
    try {
      const tableFormat = {
        tableTitle: MISAResource.table.title,
        sheetName: MISAResource.table.title,
        columns: [
          {
            fieldName: "Order",
            displayName: MISAResource.table.columnName.order,
            format: MISAEnum.CONTENT_FORMAT.INT,
            align: MISAEnum.ALIGN.MIDDLE_CENTER,
          },
          {
            fieldName: "AwardName",
            displayName: MISAResource.table.columnName.name,
            format: MISAEnum.CONTENT_FORMAT.NAME,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
          {
            fieldName: "AwardCode",
            displayName: MISAResource.table.columnName.code,
            format: MISAEnum.CONTENT_FORMAT.STRING,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
          {
            fieldName: "AwardObject",
            displayName: MISAResource.table.columnName.object,
            format: MISAEnum.CONTENT_FORMAT.ENUM,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
          {
            fieldName: "AwardLevel",
            displayName: MISAResource.table.columnName.level,
            format: MISAEnum.CONTENT_FORMAT.ENUM,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
          {
            fieldName: "AwardType",
            displayName: MISAResource.table.columnName.type,
            format: MISAEnum.CONTENT_FORMAT.ENUM,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
          {
            fieldName: "AwardStatus",
            displayName: MISAResource.table.columnName.status,
            format: MISAEnum.CONTENT_FORMAT.ENUM,
            align: MISAEnum.ALIGN.MIDDLE_LEFT,
          },
        ],
      };
      const res = await request.post(`/v1/Awards/Exportation`, tableFormat, {
        responseType: "blob",
      });
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Lấy ra file nhập khẩu mẫu
   * Created by: ntlong ( 18/07/2023 )
   */
  async getSampleImportFile() {
    try {
      const res = await request.get(`/v1/Awards/SampleImportFile`, {
        responseType: "blob",
      });
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Lấy thông tin và kiểm tra tính hợp lệ của file được upload
   * @param {File} file: File cần lấy thông tin
   * Created by: ntlong ( 25/07/2023 )
   */
  async getFileInfo(file) {
    try {
      // Tạo đối tượng FormData và thêm tệp tin vào
      const formData = new FormData();
      formData.append("file", file);
      const res = await request.post("/v1/Awards/FileInfo", formData);
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Lấy ra số bản ghi hợp lệ và không hợp lệ của file được upload
   * @param {File} file: File cần lấy thông tin
   * @param {Int} sheetIndex: Thứ tự của sheet cần check
   * @param {Int} titleLine: Thứ tự của dòng tiêu đề
   * Created by: ntlong ( 25/07/2023 )
   */
  async getImportDataStatistics(file, sheetIndex, titleLine) {
    try {
      // Tạo đối tượng FormData và thêm thông tin
      const formData = new FormData();
      formData.append("file", file);
      formData.append("sheetIndex", sheetIndex);
      formData.append("titleLine", titleLine);
      const res = await request.post("/v1/Awards/FileChecking", formData);
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },

  /**
   * Nhập dữ liệu từ file vào database
   * @param {File} file: File cần lấy thông tin
   * @param {Int} sheetIndex: Thứ tự của sheet cần check
   * @param {Int} titleLine: Thứ tự của dòng tiêu đề
   * Created by: ntlong ( 25/07/2023 )
   */
  async importExcel(file, sheetIndex, titleLine) {
    try {
      // Tạo đối tượng FormData và thêm thông tin
      const formData = new FormData();
      formData.append("file", file);
      formData.append("sheetIndex", sheetIndex);
      formData.append("titleLine", titleLine);
      const res = await request.post("/v1/Awards/Importation", formData);
      return res.data;
    } catch (error) {
      console.log(error);
    }
  },
};
